09-maze_solver/progress/internal/solver/explore.go

```diff
@@ -0,0 +1,93 @@
+package solver
+
+import (
+	"fmt"
+	"image"
+	"log/slog"
+	"slices"
+	"sync"
+)
+
+func (s *Solver) listenToBranches() {
+	wg := sync.WaitGroup{}
+	defer wg.Wait()
+
+	for {
+		select {
+		// s.quit will never return a value, unless something writes in it (which we don't do)
+		// or it has been closed, which we do when we find the treasure.
+		case <-s.quit:
+			slog.Info(fmt.Sprint("the treasure has been found, worker going to sleep"))
+			return
+		case p := <-s.pathsToExplore:
+			wg.Add(1)
+			go func(p []image.Point) {
+				defer wg.Done()
+
+				s.explore(p)
+			}(p)
+		}
+	}
+}
+
+// explore one path and publish to the s.pathsToExplore channel any branch we discover that we don't take.
+func (s *Solver) explore(pathToBranch []image.Point) {
+	// A path starts at the entrance and has stepped into the maze
+	// for at least 1 pixel, for a total of 2 pixels minimum.
+	if len(pathToBranch) < 2 {
+		return
+	}
+
+	pos := pathToBranch[len(pathToBranch)-1]
+	previous := pathToBranch[len(pathToBranch)-2]
+
+	for {
+		// Let's first check whether we should quit.
+		select {
+		case <-s.quit:
+			return
+		case s.exploredPixels <- pos:
+		}
+
+		// We know we'll have up to 3 new neighbours to explore.
+		candidates := make([]image.Point, 0, 3)
+		for _, n := range neighbours(pos) {
+			if n == previous {
+				continue
+			}
+
+			switch s.maze.RGBAAt(n.X, n.Y) {
+			case s.config.treasureColour:
+				s.solution = append(pathToBranch, n)
```

Comment:
There's a data race here when there are multiple routes to the treasure.

One approach for handling this would be to communicate the winning location via separate channel.

This also suggests that some tests are missing. One quick way to get tests for corner cases would be to generate an entirely random map using fuzzing and trying to solve it. It should be able to find both of these races.

Reply:
One of the assumptions we've made is that there is each point can be accessed via a single path (no loops). We'll see what we can do here.

Reply:
PS: the algorithm should work fine with loops, as long as the races are fixed.